enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/ArtifactUtils.java

```diff
@@ -18,70 +18,201 @@
-import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.maven.RepositoryUtils;
 import org.apache.maven.artifact.Artifact;
 import org.apache.maven.artifact.versioning.InvalidVersionSpecificationException;
 import org.apache.maven.enforcer.rule.api.EnforcerRuleException;
+import org.apache.maven.enforcer.rule.api.EnforcerRuleHelper;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.DependencyManagement;
 import org.apache.maven.plugins.enforcer.utils.ArtifactMatcher.Pattern;
-import org.apache.maven.shared.dependency.graph.DependencyNode;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluationException;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.aether.DefaultRepositorySystemSession;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.artifact.ArtifactTypeRegistry;
+import org.eclipse.aether.collection.CollectRequest;
+import org.eclipse.aether.collection.DependencyCollectionException;
+import org.eclipse.aether.collection.DependencySelector;
+import org.eclipse.aether.graph.DependencyNode;
+import org.eclipse.aether.util.graph.manager.DependencyManagerUtils;
+import org.eclipse.aether.util.graph.selector.AndDependencySelector;
+import org.eclipse.aether.util.graph.transformer.ConflictResolver;
+import org.eclipse.aether.util.graph.transformer.JavaScopeDeriver;
+import org.eclipse.aether.util.graph.transformer.JavaScopeSelector;
+import org.eclipse.aether.util.graph.transformer.NearestVersionSelector;
+import org.eclipse.aether.util.graph.transformer.SimpleOptionalitySelector;
 
 /**
  *
  * @author Robert Scholte
  * @since 3.0.0
  */
 public final class ArtifactUtils {
-    private ArtifactUtils() {}
+    private static final Map<DependencyNode, Artifact> NODE_ARTIFACT_MAP = new HashMap<>();
+    private static final Map<Artifact, org.eclipse.aether.artifact.Artifact> AETHER_ARTIFACT_MAP = new HashMap<>();
+
+    /**
+     * Converts {@link DependencyNode} to {@link Artifact}; in comparison
+     * to {@linkpl RepositoryUtils#toArtifact(org.eclipse.aether.artifact.Artifact)}, this method
+     * assigns {@link Artifact#getScope()} and {@link Artifact#isOptional()} based on
+     * the dependency information from the node.
+     *
+     * @param node {@link DependencyNode} to convert to {@link Artifact}
+     * @return target artifact
+     */
+    public static Artifact toArtifact(DependencyNode node) {
+        Artifact result = NODE_ARTIFACT_MAP.computeIfAbsent(node, n -> RepositoryUtils.toArtifact(n.getArtifact()));
+        ofNullable(node.getDependency()).ifPresent(dependency -> {
+            ofNullable(dependency.getScope()).ifPresent(result::setScope);
+            result.setOptional(dependency.isOptional());
+        });
```

Comment:
This will be computed always and we change item which is in `NODE_ARTIFACT_MAP` ...
So result in multithread can be corrupted by other thread - we have the same instance.

Reply:
Agreed.

Reply:
Updated: caching the built objects.